# Last Index Of Number Question

Given an array of length N and an integer x, you need to find and return the last index of integer x present in the array.<br> Return -1 if it is not present in the array.<br>
Last index means - if x is present multiple times in the array, return the index at which x comes last in the array.<br>
You should start traversing your array from 0, not from (N - 1).<br>
Do this recursively. Indexing in the array starts from 0.<br>

In [1]:
def lastIndex(arr, x, si):
    l = len(arr)
    if si == l:
        return -1
   
    smallerlistoutput = lastIndex(arr,x,si+1)
    if smallerlistoutput != -1:
        return smallerlistoutput
    else:
        if arr[si] == x:
            return si
        else:
            return -1
    
    
# Main
from sys import setrecursionlimit
setrecursionlimit(11000)
n=int(input())
arr=list(int(i) for i in input().strip().split(' '))
x=int(input())
print(lastIndex(arr, x, 0))


4
9 8 10 8
8
3


# Remove X

Given a string, compute recursively a new string where all 'x' chars have been removed.

In [2]:
def removeX(string):
    if len(string) == 0:
        return string
    
    smalloutput = removeX(string[1:])
    if string[0] == 'x':
        return smalloutput
    else:
        return string[0] + smalloutput

# Main
string = input()
print(removeX(string))


xaxb
ab


# Remove Duplicates Recursively

Given a string S, remove consecutive duplicates from it recursively.

In [3]:
def removeConsecutiveDuplicates(string):
    
    l = len(string)
    if l == 0 or l == 1:
        return string
    if string[0] == string[1]:
        smalloutput = removeConsecutiveDuplicates(string[1:])
        return smalloutput
    else:
        smalloutput = removeConsecutiveDuplicates(string[1:])
        return  string[0]+smalloutput

# Main
string = input().strip()
print(removeConsecutiveDuplicates(string))


aabccba
abcba


# Tower Of Hanoi - Problem

Tower of Hanoi is a mathematical puzzle where we have three rods and n disks. The objective of the puzzle is to move all <br>
disks from source rod to destination rod using third rod (say auxiliary). The rules are :<br>
1) Only one disk can be moved at a time.<br>
2) A disk can be moved only if it is on the top of a rod.<br>
3) No disk can be placed on the top of a smaller disk.<br>
Print the steps required to move n disks from source rod to destination rod.<br>
Source Rod is named as 'a', auxiliary rod as 'b' and destination rod as 'c'.<br>

In [4]:
def towerofhanoi(n, source, aux, dest):
    if n == 0:
        return
    if n == 1:
        print(source, dest)
        return
    towerofhanoi(n-1,source,dest,aux)
    print(source,dest)
    towerofhanoi(n-1,aux,source,dest)
    

n=int(input())
towerofhanoi(n, 'a', 'b', 'c')


3
a c
a b
c b
a c
b a
b c
a c


# Geometric Sum

Given k, find the geometric sum i.e.
1 + 1/2 + 1/4 + 1/8 + ... + 1/(2^k) 

In [5]:
## Read input as specified in the question.
## Print output as specified in the question.
def findSum(k):
    if k == 0:
        return 1
    sumoutput = findSum(k-1)
    a = sumoutput + (1/(2**k))
    return float(a)

k = int(input())
x = findSum(k)
print('%.5f'%x)

4
1.93750


# Check Palindrome (recursive)

Check whether a given String S is a palindrome using recursion. Return true or false.

In [6]:
## Read input as specified in the question.
## Print output as specified in the question.
def palindrome(s,si,ei):
    
    if ei-si == 1 or ei-si == 0:
        return True
        
    
    if s[si]==s[ei]:
        return palindrome(s,si+1,ei-1)    
    else:
        return False
s=input()
ans = palindrome(s,0,len(s)-1)
if ans:
    print('true')
else:
    print('false')
    


racecar
true


# Sum of digits (recursive)

Write a recursive function that returns the sum of the digits of a given integer.

In [7]:
## Read input as specified in the question.
## Print output as specified in the question.
def SumofDigits(n):
    if n == 0:
        return 0
    sum = 0
    n1 = n%10
    n = n//10
    sumoutput = SumofDigits(n)
    return sum + sumoutput + n1

n = int(input())
print(SumofDigits(n))

12345
15


# Count Zeros

Given an integer N, count and return the number of zeros that are present in the given integer using recursion

In [8]:
## Read input as specified in the question.
## Print output as specified in the question.
def CountZeros(n):
    if n<10:
        if n == 0:
            return 1
        return 0
    
    sum = 0
    if n % 10 == 0:
        sum = sum + 1
    n1 = n // 10
    smalloutput = sum + CountZeros(n1)
    return smalloutput  


n = int(input())
x = CountZeros(n)
print(x)

00010204
2


Even though "00010204" has 5 zeros, the output would still be 2 because when you convert it to an integer, it becomes 10204.

# String to Integer

Write a recursive function to convert a given string into the number it represents. That is input will be a numeric string <br>
that contains only numbers, you need to convert the string into corresponding integer and return the answer.<br>

In [9]:
## Read input as specified in the question.
## Print output as specified in the question.
def StoI(string,start,end):
    n = end - start 
    a = int(string[start])
    if n == 0 :
        
        a * 10**(n) 
        return a * 10**(n) 
    smalleroutput = StoI(string,start+1,end)    
    return int(string[start])*10**(n) +  smalleroutput 


string = input()
x = StoI(string,0,len(string)-1)
print(x)

00001231
1231


# Pair Star

Given a string S, compute recursively a new string where identical chars that are adjacent in the original string are <br>
separated from each other by a "*".<br>

In [10]:
## Read input as specified in the question.
## Print output as specified in the question.
def PairStart(string,start,end):
    if end - start == 0:
        return string[start]
    smalloutput = PairStart(string,start+1,end)
    if string[start] == string[start+1]:
        return  string[start:start+1] + '*' + smalloutput
    else:
        return string[start:start+1]  + smalloutput



string = input()
x = PairStart(string,0,len(string)-1)
print(x)

hello
hel*lo


# Check AB

Suppose you have a string, S, made up of only 'a's and 'b's. Write a recursive function that checks if the string was <br>
generated using the following rules:<br>
a. The string begins with an 'a'<br>
b. Each 'a' is followed by nothing or an 'a' or "bb"<br>
c. Each "bb" is followed by nothing or an 'a'<br>
If all the rules are followed by the given string, return true otherwise return false.<br>

In [11]:
## Read input as specified in the question.
## Print output as specified in the question.
def CheckAB(str):
    if len(str)==0:
        return True
    if str[0]=="a":
        if len(str[1:])>1 and str[1:3]=='bb':
            return CheckAB(str[3:])
        else:
            return CheckAB(str[1:])
    else:
        return False

s = input()
if CheckAB(s) is True:
    print('true')
else:
    print('false')

abababa
false


# Staircase

A child is running up a staircase with N steps, and can hop either 1 step, 2 steps or 3 steps at a time. Implement a <br>
method to count how many possible ways the child can run up to the stairs. You need to return number of possible ways W.<br>

In [12]:
## Read input as specified in the question.
## Print output as specified in the question.
def Staircase(n):
    if n == 0 or n == 1: 
        return 1
    elif n == 2:
        return 2
    return Staircase(n-1) + Staircase(n-2) + Staircase(n-3)

n = int(input())
x = Staircase(n)
print(x)

4
7


# Return Subsequences

Given a string (let's say of length n), return all the subsequences of the given string.<br>
Subsequences contain all the strings of length varying from 0 to n. But the order of characters should remain the same as<br> 
in the input string.<br>
Note: The order of subsequences are not important.<br>

In [13]:
def subsequences(string):

    if len(string) == 0:
        output = []
        output.append(" ")
        return output
    
    smallerString = string[1:]
    smallerOutput = subsequences(smallerString)
    
    output = []
    for sub in smallerOutput:
        output.append(sub)
        
    for sub in smallerOutput:
        subs_with_zeroth_char = string[0] + sub
        output.append(subs_with_zeroth_char)
    
    return output

string = input()
ans = subsequences(string)
for ele in ans:
    print(ele)

abc
 
c 
b 
bc 
a 
ac 
ab 
abc 


# Return Keypad

Given an integer n, using phone keypad find out and return all the possible strings that can be made using digits of <br>
input n.<br>
Note : The order of strings are not important.<br>
Input Format :<br>
Integer n<br>

In [14]:
def getString(d):
    if d == 2:
        return "abc"
    if d == 3:
        return "def"
    if d == 4:
        return "ghi"
    if d == 5:
        return "jkl"
    if d == 6:
        return "mno"
    if d == 7:
        return "pqrs"
    if d == 8:
        return "tuv"
    if d == 9:
        return "wxyz"
    return " "   
    

def keypad(input):
    if input == 0:
        output = []
        output.append("")
        return output
    
    smallerInput = input//10
    remainder = input%10    
    
    smallerOutput = keypad(smallerInput)
    text = getString(remainder)
        
    output = []
    for s in smallerOutput:
        for c in text:
            option = s + c
            output.append(option)
            
    return output

input = int(input())
ans = keypad(input)
for s in ans:
    print(s)

23
ad
ae
af
bd
be
bf
cd
ce
cf


# Print Subsequences

Given a string (lets say of length n), print all the subsequences of the given string.<br>
Subsequences contain all the strings of length varying from 0 to n. But the order of characters should remain same as in <br>
the input string.<br>
Note :<br>
The order of subsequences are not important. Print every subsequence in new line.<br>

In [2]:
def printSub(str, output):
    if len(str) == 0:
        print(output)
        return
    
    # Not including 0th char
    printSub(str[1:], output)
    
    # Including 0th char
    printSub(str[1:], output+str[0])

str = input()
printSub(str, "")

abc

c
b
bc
a
ac
ab
abc


# Return subsets sum to K

Given an array A of size n and an integer K, return all subsets of A which sum to K.<br>
Subsets are of length varying from 0 to n, that contain elements of the array. But the order of elements should <br>
remain same as in the input array.<br>
Note : The order of subsets are not important.<br>
Input format :<br>
Line 1 : Integer n, Size of input array<br>
Line 2 : Array elements separated by space<br>
Line 3 : K <br>

In [3]:
import sys
sys.setrecursionlimit(10 ** 8)


def subs(str):
    if len(str) == 0:
        output = []
        output.append(" ")
        return output
    
    smallerString = str[1:]
    smallerOutput = subs(smallerString)
    
    output = []
    for sub in smallerOutput:
        output.append(sub)
        
    for sub in smallerOutput:
        subs_with_zeroth_char = str[0] + sub
        output.append(subs_with_zeroth_char)
    
    return output
    
def subsetsSumK(arr, k) :
    #base cases
    n=len(arr)
    if n==1:
        if arr[0]==k:
            output=[[k]]
            return output
        else:
            output=[]
            return output
#recursive callsss
    out1=subsetsSumK(arr[:-1],k)
    out2=subsetsSumK(arr[:-1],k-arr[-1])
    for lst in out2:
        lst.append(arr[-1])
    if arr[-1]==k:
        out2.append([k])
    return out1+out2

#taking input
def takeInput() :
    n = int(input().strip())

    if n == 0 :
        return list(), 0

    arr = [int(element) for element in list(input().strip().split(" "))]
    return arr, n


#printing the list of lists
def printListOfList(liOfLi) :
    for li in liOfLi :
        for elem in li :
            print(elem, end = " ")
        print()

#main
arr, n = takeInput()

if n != 0 :
    k = int(input().strip())
    liOfLi = subsetsSumK(arr, k)

    printListOfList(liOfLi)



9
5 12 3 17 1 18 15 3 17 
6
5 1 
3 3 


# Return subset of an array

Given an integer array (of length n), find and return all the subsets of input array.<br>
Subsets are of length varying from 0 to n, that contain elements of the array. But the order of elements should remain <br>
same as in the input array.<br>
Note : The order of subsets are not important.<br>
Input format :<br>

Line 1 : Size of array<br>

Line 2 : Array elements (separated by space)<br>

In [4]:
def subs(l):
    if l == []:
        return [[]]

    x = subs(l[1:])

    return x + [[l[0]] + y for y in x]

n = int(input())
l = [int(element) for element in list(input().strip().split(" "))]

x = subs(l)
for i in x:
    for j in i:
        print(j, end = " ")
    print()

3
15 20 12

12 
20 
20 12 
15 
15 12 
15 20 
15 20 12 


# Return Permutations of a String

Given a string, find and return all the possible permutations of the input string.<br>
Note : The order of permutations are not important.

In [5]:
def permutations(s):
 
    # create a list to store (partial) permutations
    partial = []
 
    # initialize the list with the first character of the string
    partial.append(s[0])
 
    # do for every character of the specified string
    for i in range(1, len(s)):
 
        # consider previously constructed partial permutation one by one
 
        # iterate backward
        for j in reversed(range(len(partial))):
 
            # remove the current partial permutation from the list
            curr = partial.pop(j)
 
            # Insert the next character of the specified string into all
            # possible positions of current partial permutation.
            # Then insert each of these newly constructed strings into the list
 
            for k in range(len(curr) + 1):
                partial.append(curr[:k] + s[i] + curr[k:])
 
    for ele in partial:
        print(ele)

s = input()
ans = permutations(s)

abc
cab
acb
abc
cba
bca
bac


# Print Subsets of Array

Given an integer array (of length n), find and print all the subsets of input array.<br>
Subsets are of length varying from 0 to n, that contain elements of the array. But the order of elements should remain <br>
same as in the input array.<br>
Note : The order of subsets are not important. Just print the subsets in different lines.<br>
Input format :<br>
Line 1 : Integer n, Size of array<br>
Line 2 : Array elements (separated by space)<br>

In [6]:
def subs(l):
    if l == []:
        return [[]]

    x = subs(l[1:])

    return x + [[l[0]] + y for y in x]

n = int(input())
l = [int(element) for element in list(input().strip().split(" "))]

x = subs(l)
for i in x:
    for j in i:
        print(j, end = " ")
    print()
    

3
15 20 12

12 
20 
20 12 
15 
15 12 
15 20 
15 20 12 


# Print Subset Sum to K

Given an array A and an integer K, print all subsets of A which sum to K.<br>
Subsets are of length varying from 0 to n, that contain elements of the array. But the order of elements should remain <br>
same as in the input array.<br>
Note : The order of subsets are not important. Just print them in different lines.<br>
Input format :<br>
Line 1 : Size of input array<br>
Line 2 : Array elements separated by space<br>
Line 3 : K <br>

In [7]:
## Read input as specified in the question.
## Print output as specified in the question.
import sys
sys.setrecursionlimit(10 ** 8)


def subs(str):
    if len(str) == 0:
        output = []
        output.append(" ")
        return output
    
    smallerString = str[1:]
    smallerOutput = subs(smallerString)
    
    output = []
    for sub in smallerOutput:
        output.append(sub)
        
    for sub in smallerOutput:
        subs_with_zeroth_char = str[0] + sub
        output.append(subs_with_zeroth_char)
    
    return output
    
def subsetsSumK(arr, k) :
    #base cases
    n=len(arr)
    if n==1:
        if arr[0]==k:
            output=[[k]]
            return output
        else:
            output=[]
            return output
#recursive callsss
    out1=subsetsSumK(arr[:-1],k)
    out2=subsetsSumK(arr[:-1],k-arr[-1])
    for lst in out2:
        lst.append(arr[-1])
    if arr[-1]==k:
        out2.append([k])
    return out1+out2

#taking input
def takeInput() :
    n = int(input().strip())

    if n == 0 :
        return list(), 0

    arr = [int(element) for element in list(input().strip().split(" "))]
    return arr, n


#printing the list of lists
def printListOfList(liOfLi) :
    for li in liOfLi :
        for elem in li :
            print(elem, end = " ")
        print()

#main
arr, n = takeInput()

if n != 0 :
    k = int(input().strip())
    liOfLi = subsetsSumK(arr, k)

    printListOfList(liOfLi)


9
5 12 3 17 1 18 15 3 17 
6
5 1 
3 3 


# Print Permutations of a String

Given a string, find and print all the possible permutations of the input string.<br>
Note : The order of permutations are not important. Just print them in different lines.<br>

In [8]:
def permutations(s):
 
    # create a list to store (partial) permutations
    partial = []
 
    # initialize the list with the first character of the string
    partial.append(s[0])
 
    # do for every character of the specified string
    for i in range(1, len(s)):
 
        # consider previously constructed partial permutation one by one
 
        # iterate backward
        for j in reversed(range(len(partial))):
 
            # remove the current partial permutation from the list
            curr = partial.pop(j)
 
            # Insert the next character of the specified string into all
            # possible positions of current partial permutation.
            # Then insert each of these newly constructed strings into the list
 
            for k in range(len(curr) + 1):
                partial.append(curr[:k] + s[i] + curr[k:])
 
    for ele in partial:
        print(ele)

s = input()
ans = permutations(s)

abc
cab
acb
abc
cba
bca
bac
